File: src/App/backend/src/Altinn.App.Core/Constants/AltinnEnvironments.cs
using System.Collections.Frozen;
using System.Globalization;
using Microsoft.Extensions.Hosting;

namespace Altinn.App.Core.Constants;

internal static class AltinnEnvironments
{
    public static IReadOnlyDictionary<HostingEnvironment, IEnumerable<string>> Map { get; } =
        new Dictionary<HostingEnvironment, IEnumerable<string>>
        {
            [HostingEnvironment.Development] =
            [
                Environments.Development.ToLower(CultureInfo.InvariantCulture),
                "dev",
                "local",
                "localtest",
            ],
            [HostingEnvironment.Staging] =
            [
                Environments.Staging.ToLower(CultureInfo.InvariantCulture),
                "test",
                "at22",
                "at23",
                "at24",
                "tt02",
                "yt01",
            ],
            [HostingEnvironment.Production] =
            [
                Environments.Production.ToLower(CultureInfo.InvariantCulture),
                "prod",
                "produksjon",
            ],
        }.ToFrozenDictionary();

    public static HostingEnvironment GetHostingEnvironment(string environmentName)
    {
        var envNameLower = environmentName.ToLower(CultureInfo.InvariantCulture);
        return Map.FirstOrDefault(x => x.Value.Contains(envNameLower)).Key;
    }

    public static HostingEnvironment GetHostingEnvironment(IHostEnvironment hostEnvironment)
    {
        return GetHostingEnvironment(hostEnvironment.EnvironmentName);
    }
}

/// <summary>
/// The kind of environment an app is hosted in, resolved from the host environment name (for example
/// <c>tt02</c> and <c>at22</c> are staging environments). Configuration that varies per environment is
/// resolved against this.
/// </summary>
public enum HostingEnvironment
{
    /// <summary>The environment name matched no known environment.</summary>
    Unknown,

    /// <summary>Local development, including localtest.</summary>
    Development,

    /// <summary>Production.</summary>
    Production,

    /// <summary>A test environment (tt02, at2x, yt01).</summary>
    Staging,
}

File: src/Designer/frontend/packages/process-editor/src/components/ConfigPanel/EnvironmentConfig/altinnEnvironments.ts
/**
 * Mirrors `Altinn.App.Core/Constants/AltinnEnvironments.cs`: the runtime folds every `env`
 * attribute into one of three environments, case-insensitively, and treats anything else as
 * unknown. Keep the two tables in sync.
 */
export type AltinnEnvironment = 'development' | 'staging' | 'production';

export const altinnEnvironments: readonly AltinnEnvironment[] = [
  'development',
  'staging',
  'production',
];

const environmentAliases: Readonly<Record<AltinnEnvironment, readonly string[]>> = {
  development: ['development', 'dev', 'local', 'localtest'],
  staging: ['staging', 'test', 'at22', 'at23', 'at24', 'tt02', 'yt01'],
  production: ['production', 'prod', 'produksjon'],
};

export const normalizeAltinnEnvironment = (env: string): AltinnEnvironment | undefined => {
  const lowerCasedEnv = env.toLowerCase();
  return altinnEnvironments.find((environment) =>
    environmentAliases[environment].includes(lowerCasedEnv),
  );
};

/** The runtime reads a missing or blank `env` as the entry that applies to every environment. */
export const isGlobalEnv = (env: string | undefined): boolean => !env?.trim();
